resources/type-system/upper-lower-bounds.md

```diff
@@ -81,6 +85,34 @@ We define the upper bound of two types T1 and T2 to be **UP**(`T1`,`T2`) as foll
 - **UP**(`T1`, `T2`) = `T2` if **BOTTOM**(`T1`)
 - **UP**(`T1`, `T2`) = `T1` if **BOTTOM**(`T2`)
 
+- **UP**(`X1 extends B1`, `T2`) =
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1a`, `T2`)
+    where `B1a` is the greatest closure of `B1` with respect to `X1`,
```

Comment:
(Why do we need the greatest closure? Is it to avoid bad recursion?)

Reply:
The argument was that we'd need to obtain a type that doesn't depend on `X`, and hence it would be a problem if `B1` is an F-bound that contains `X`.

Reply:
ACK.